File: ms-inventario/src/main/java/com/smartlogix/ms_inventario/controller/ProductoController.java
package com.smartlogix.ms_inventario.controller;

import com.smartlogix.ms_inventario.dto.ProductoDTO;
import com.smartlogix.ms_inventario.model.Producto;
import com.smartlogix.ms_inventario.service.ProductoService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/productos")
public class ProductoController {

    @Autowired
    private ProductoService service;

    @GetMapping
    public List<Producto> listarTodos() {
        return service.listarTodos();
    }

    @GetMapping("/bajo-stock")
    public List<Producto> listarBajoStock() {
        return service.listarBajoStock();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Producto> obtenerPorId(@PathVariable Long id) {
        return ResponseEntity.ok(service.obtenerPorId(id));
    }

    @PostMapping
    public ResponseEntity<Producto> crearProducto(@Valid @RequestBody ProductoDTO dto) {
        return new ResponseEntity<>(service.crearProducto(dto), HttpStatus.CREATED);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Producto> actualizarProducto(@PathVariable Long id, @Valid @RequestBody ProductoDTO dto) {
        return ResponseEntity.ok(service.actualizarProducto(id, dto));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> eliminarProducto(@PathVariable Long id) {
        service.eliminarProducto(id);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{id}/stock")
    public ResponseEntity<Producto> ajustarStock(@PathVariable Long id, @RequestParam int cantidad) {
        return ResponseEntity.ok(service.ajustarStock(id, cantidad));
    }
}

File: ms-inventario/src/productos/producto.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Patch,
  Body,
  Param,
  Query,
  HttpCode,
  HttpStatus,
  ParseIntPipe,
} from '@nestjs/common';
import { ProductoService } from './producto.service';
import { ProductoDTO } from './producto.dto';
import { Producto } from './producto.entity';

@Controller('productos')
export class ProductoController {
  constructor(private readonly service: ProductoService) {}

  @Get()
  async listarTodos(): Promise<Producto[]> {
    return this.service.listarTodos();
  }

  @Get('bajo-stock')
  async listarBajoStock(): Promise<Producto[]> {
    return this.service.listarBajoStock();
  }

  @Get(':id')
  async obtenerPorId(@Param('id', ParseIntPipe) id: number): Promise<Producto> {
    return this.service.obtenerPorId(id);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async crearProducto(@Body() dto: ProductoDTO): Promise<Producto> {
    return this.service.crearProducto(dto);
  }

  @Put(':id')
  async actualizarProducto(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: ProductoDTO,
  ): Promise<Producto> {
    return this.service.actualizarProducto(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async eliminarProducto(@Param('id', ParseIntPipe) id: number): Promise<void> {
    await this.service.eliminarProducto(id);
  }

  @Patch(':id/stock')
  async ajustarStock(
    @Param('id', ParseIntPipe) id: number,
    @Query('cantidad', ParseIntPipe) cantidad: number,
  ): Promise<Producto> {
    return this.service.ajustarStock(id, cantidad);
  }
}
